Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp

```diff
@@ -2190,7 +2190,6 @@ void MediaPlayerPrivateGStreamer::updateBufferingStatus(GstBufferingMode mode, d
     GST_DEBUG_OBJECT(pipeline(), "[Buffering] mode: %s, status: %f%%", enumToString(GST_TYPE_BUFFERING_MODE, mode).data(), percentage);
 
     m_didDownloadFinish = percentage == 100;
-    m_isBuffering = !m_didDownloadFinish;
```

Comment:
The removal of this line is problematic. It's making the [video-hls-copy-into-canvas.html](https://github.com/WebKit/WebKit/blob/main/LayoutTests/http/tests/media/video-hls-copy-into-canvas.html) test fail upstream.

That test [uses a GST_BUFFERING_STREAM mode](https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/wpe-2.28/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp#L2209). What's happening there is that `m_isBuffering` never goes back from true to false, and therefore [the readyState never reaches HaveEnoughData following the m_didDownloadFinish path](https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/wpe-2.28/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp#L2502).

Instead, it keeps going the [m_isBuffering path](https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/wpe-2.28/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp#L2483), [queries again the m_idBuffering status](https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/wpe-2.28/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp#L2487), which keeps being true, and enters [to the branch that sets HasCurrentData](https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/wpe-2.28/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp#L2498).

To have this issue fixed, the [buffering correction code](https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/wpe-2.28/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp#L2486) would need a couple of changes. It was a bit fishy anyway that it considered "buffering" as `m_bufferingPercentage == 100`. To my understanding, it should be the opposite!

Please, could you check if [this change](https://github.com/eocanha/WebKit/commit/01bdc3fe99f46341e21a22eb86d5a90f5363c295) from my [metro-pr945-debug](https://github.com/eocanha/WebKit/commits/metro-pr945-debug) branch (based on **upstream**!) still keeps the fix of your original commit working for your use case?

On my side I still have to check if that added fix causes any other regression upstream.

Reply:
Hi, the change for "m_bufferingPercentage == 100" doesn't really matter in my case as m_buffering will be updated from pipeline querry anyway. Does this pipeline querry return that buffering is sill in progress in your case or the querry  fails leaving m_isBuffering = m_bufferingPercentage == 100?

But even with "m_isBuffering = m_bufferingPercentage < 100; // ### DEBUG" fixed the change to remove " m_isBuffering = !m_didDownloadFinish;" is still needed. Otherwise when buffering was finished from queue message the m_isBuffering will be silently changed to false so updateStates() will never recognize it as wasBuffering and now is not. Btw according to git blame this line https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/wpe-2.28/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp#L2483 was previously based on buffering percentage var (or download finished, not sure now) so it was making much more sense

Reply:
Quick reply to one of the questions:

> But even with "m_isBuffering = m_bufferingPercentage < 100; // ### DEBUG" fixed the change to remove " m_isBuffering = !m_didDownloadFinish;" is still needed.

Yes, it's fine that the line removal is still needed. The `m_bufferingPercentage < 100` change I'm proposing is to fix things with the failing HLS test when the line is removed.

Reply:
sure, that  is perfectly fine for me, thanks

Reply:
Great. It seems that removing the line and adding the `< 100` change there are no regressions. I'm working on having those changes approved and landed upstream.